tools/clang/lib/SPIRV/SPIRVEmitter.cpp

```diff
@@ -1345,6 +1336,147 @@ uint32_t SPIRVEmitter::doConditionalOperator(const ConditionalOperator *expr) {
+  if (TypeTranslator::isStructuredBuffer(type)) {
+    uint32_t size = 0, stride = 0;
+    std::tie(std::ignore, size) = typeTranslator.getAlignmentAndSize(
+        type, LayoutRule::GLSLStd430, /*isRowMajor*/ false, &stride);
+    const auto sizeId = theBuilder.getConstantUint32(size);
+    theBuilder.createStore(doExpr(expr->getArg(1)), sizeId);
+  }
+
+  return 0;
+}
+
+uint32_t
+SPIRVEmitter::processBufferTextureGetDimensions(const CXXMemberCallExpr *expr) {
+  theBuilder.requireCapability(spv::Capability::ImageQuery);
+  const auto *object = expr->getImplicitObjectArgument();
+  const auto objectId = loadIfGLValue(object);
+  const auto type = object->getType();
+  const auto *recType = type->getAs<RecordType>();
+  assert(recType);
+  const auto typeName = recType->getDecl()->getName();
+  const auto numArgs = expr->getNumArgs();
+  const Expr *mipLevel = nullptr, *numLevels = nullptr, *numSamples = nullptr;
+
+  assert(TypeTranslator::isTexture(type) || TypeTranslator::isRWTexture(type) ||
+         TypeTranslator::isBuffer(type) || TypeTranslator::isRWBuffer(type));
+
+  // For Texture1D, arguments are either:
+  // a) width
+  // b) MipLevel, width, NumLevels
+
+  // For Texture1DArray, arguments are either:
+  // a) width, elements
+  // b) MipLevel, width, elements, NumLevels
+
+  // For Texture2D, arguments are either:
+  // a) width, height
+  // b) MipLevel, width, height, NumLevels
+
+  // For Texture2DArray, arguments are either:
+  // a) width, height, elements
+  // b) MipLevel, width, height, elements, NumLevels
+
+  // For Texture3D, arguments are either:
+  // a) width, height, depth
+  // b) MipLevel, width, height, depth, NumLevels
+
+  // For Texture2DMS, arguments are: width, height, NumSamples
+
+  // For Texture2DMSArray, arguments are: width, height, elements, NumSamples
+
+  if ((typeName == "Texture1D" && numArgs > 1) ||
+      (typeName == "Texture2D" && numArgs > 2) ||
+      (typeName == "Texture3D" && numArgs > 3) ||
+      (typeName == "Texture1DArray" && numArgs > 2) ||
+      (typeName == "Texture2DArray" && numArgs > 3)) {
+    mipLevel = expr->getArg(0);
+    numLevels = expr->getArg(numArgs - 1);
+  }
+  if (TypeTranslator::isTextureMS(type)) {
+    numSamples = expr->getArg(numArgs - 1);
+  }
+
+  uint32_t querySize = numArgs;
+  // If numLevels arg is present, mipLevel must also be present. These are not
+  // queried via ImageQuerySizeLod.
+  if (numLevels)
+    querySize -= 2;
+  // If numLevels arg is present, mipLevel must also be present.
+  else if (numSamples)
+    querySize -= 1;
+
+  const uint32_t uintId = theBuilder.getUint32Type();
+  const uint32_t resultTypeId =
+      querySize == 1 ? uintId : theBuilder.getVecType(uintId, querySize);
+
+  // Only Texture types use ImageQuerySizeLod.
+  // TextureMS, RWTexture, Buffers, RWBuffers use ImageQuerySize.
+  uint32_t lod = 0;
+  if (TypeTranslator::isTexture(type) && !numSamples) {
```

Comment:
I think we can have an additional `bool` named as like `isNonMSTexture` defined and turn it into true at line 1413. So we do not need to call `isTexture()` again.

Reply:
Ah, that's right. Sorry.